### PR TITLE
Fix SPI and SD for BIGTREETECH SKR Mini V1.1

### PR DESCRIPTION
**Pull request by 3DSmitty**
### Description
This change fixes the SPI code for STM32F1 boards especially the BIGTREETECH SKR Mini V1.1 board. It also allows you to choose which SPI device to use for your application (SPI1, SPI2, or SPI3). On the BIGTREETECH SKR Mini V1.1 board the onboard SD card reader is connected to SPI1 and the external SD card reader (On LCD) is connected to SPI3. You are able to choose which one you would like to use from pins_BIGTREE_SKR_MINI_V1_1.h file.

### Benefits
Allows use of the onboard SD card reader or the external SD card reader. 

### Related Issues
This is related to https://github.com/MarlinFirmware/Marlin/pull/14041 to add Marlin support for the BIGTREETECH SKR Mini V1.1

### CODE REVIEW

**Comment by tpruvot** [comment]
did you see you can use 	beginTransaction(BOARD_SPI_DEFAULT_SS, settings);
https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/libraries/SPI/src/SPI.h#L217

may avoid to fully break TMC and eeprom SPI

**Comment by gloomyandy** [comment]
I see that this code is manipulating SS_PIN on each side of a read/write operation. My understanding is that other HALs do not do this and leave the handling of chip select pins to external code. I believe that this is so that chip select can be held over multiple SPI operations. So what is the correct way to handle the SS_PIN? @p3p @thinkyhead

**Comment by 3DSmitty** [comment]
> did you see you can use 	beginTransaction(BOARD_SPI_DEFAULT_SS, settings);
> https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/libraries/SPI/src/SPI.h#L217
> may avoid to fully break TMC and eeprom SPI

@tpruvot yes I did see that but I don't see where it does anything with the SS_PIN after that. I tried to use it with no results.

This should not effect the EEPROM SPI

**Comment by tpruvot** [comment]
Indeed... 🤦‍♂

**Comment by tpruvot** [comment]
eeprom spi code may need some work...

Marlin/src/HAL/shared/SpiEeprom.cpp
```cpp
uint8_t eeprom_read_byte(uint8_t* pos) {
  uint8_t v;
  uint8_t eeprom_temp[3];

  // set read location
  // begin transmission from device
  eeprom_temp[0] = CMD_READ;
  eeprom_temp[1] = ((unsigned)pos>>8) & 0xFF; // addr High
  eeprom_temp[2] = (unsigned)pos& 0xFF;       // addr Low
  WRITE(SPI_EEPROM1_CS, HIGH);
  WRITE(SPI_EEPROM1_CS, LOW);
  **spiSend(SPI_CHAN_EEPROM1, eeprom_temp, 3);**

  v = **spiRec(SPI_CHAN_EEPROM1);**
  WRITE(SPI_EEPROM1_CS, HIGH);
  return v;
}
```

**Comment by thisiskeithb** [comment]
Can we update the ascii art for the EXP headers while we're here?

```
/*
|                 _____                                             _____
|             NC | · · | GND                                    5V | · · | GND
|          RESET | · · | PB9 (SD_DETECT)             (LCD_D7) PC14 | · · | PC15 (LCD_D6)
|   (MOSI)   PB5 | · · | PB8 (BTN_EN2)               (LCD_D5)  PB7 | · · | PC13 (LCD_D4)
|  (SD_SS)  PA15 | · · | PD2 (BTN_EN1)               (LCD_RS) PC12 | · · | PB6  (LCD_EN)
|    (SCK)   PB3 | · · | PB4 (MISO)                 (BTN_ENC) PC11 | · · | PC10 (BEEPER)
|                 ￣￣                                               ￣￣  
|                 EXP2                                              EXP1  
*/
```

**Comment by 3DSmitty** [comment]
> eeprom spi code may need some work...

Wow that's a lot different than done in the ststm32 library!

**Review by 3DSmitty — COMMENTED** [review]
I like it!

**Comment by tpruvot** [comment]
yep, anyway for the eeprom, the functions with the channel parameter could be tuned to avoid the SS_PIN switch

File: Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h [hunk 1]
```diff
@@ -123,10 +126,13 @@
...
+// By default the onboard SD is enabled. 
+// To disable it and use an external SD (connected to LCD)
+// enable STM32_SD_LCD.
+
 //#define STM32_SD_LCD
 
-#ifdef STM32_SD_LCD
+#if ENABLED(STM32_SD_LCD)
   #define SD_DETECT_PIN    PB9
   #define ENABLE_SPI3
```
Comment (Msq001) [inline Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h:137]: #define ENABLE_SPI3 
  should be 
  #define DISABLE_JTAG

Comment (thinkyhead) [reply Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h:137]: Since that is already always the case, I guess the line may be removed entirely.

**Comment by thinkyhead** [comment]
> `#define ENABLE_SPI3` [can be removed]

> …the functions with the channel parameter could be tuned to avoid the `SS_PIN` switch…

Once these changes are made and everything tests out well, we can merge this!

**Comment by gloomyandy** [comment]
Folks what is going on here? Why does the stm HAL need to manipulate the SS_PIN in the read and write functions when other HALs do not do this? They can't both be correct. Surely by having these functions set this pin low/high it means that you can only use the HAL SPI functions to talk to a single device? Am I missing something?

**Comment by tpruvot** [comment]
yes, well, these simplified spi functions are for the sdcard on marlin, others (dac & eeprom) use another prototype with the channel and were just added in the HAL... so nothing to break here

**Comment by p3p** [comment]
@gloomyandy You are correct, the SPI functions in the HAL should not interfere it is up to the caller to correctly select the target device.

File: Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h (hunk 1 above)
Comment (3DSmitty) [reply Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h:137]: > #define ENABLE_SPI3
> should be
> #define DISABLE_JTAG

@Msq001 Why do you say that? `DISABLE_JTAG` is already defined at the top of `pins_BIGTREE_SKR_MINI_V1_1.h`. `ENABLE_SPI3` determines which SPI device to use and which set of SPI pins to use in `\HAL_STM32F1\spi_pins.h`
```cpp
#if ENABLED(ENABLE_SPI3)
  #define SPI_DEVICE 3
  #define SCK_PIN   BOARD_SPI3_SCK_PIN
  #define MISO_PIN  BOARD_SPI3_MISO_PIN
  #define MOSI_PIN  BOARD_SPI3_MOSI_PIN
  #define SS_PIN    BOARD_SPI3_NSS_PIN
#elif ENABLED(ENABLE_SPI2)
  #define SPI_DEVICE 2
  #define SCK_PIN   BOARD_SPI2_SCK_PIN
  #define MISO_PIN  BOARD_SPI2_MISO_PIN
  #define MOSI_PIN  BOARD_SPI2_MOSI_PIN
  #define SS_PIN    BOARD_SPI2_NSS_PIN
#else
  #define SPI_DEVICE 1
  #define SCK_PIN   BOARD_SPI1_SCK_PIN
  #define MISO_PIN  BOARD_SPI1_MISO_PIN
  #define MOSI_PIN  BOARD_SPI1_MOSI_PIN
  #define SS_PIN    BOARD_SPI1_NSS_PIN
#endif
```

**Comment by 3DSmitty** [comment]
@gloomyandy there are a few different ways to do SPI. I don't know if there is a "correct" way or not. The previous way was not working for reading SD cards in Marlin. Manipulating the SS_PIN does work for reading and writing to SD cards. With this pull request you should be able to enable more than one SPI Master device through spi_pins.h. By using the SPI.setModule(SPI_DEVICE) function you can switch which SPI device you are using.

**Comment by 3DSmitty** [comment]
It would be great if we could get @rogerclarkmelbourne feedback on this pull request

**Comment by gloomyandy** [comment]
Honestly this really does not make any sense to me. The Marlin SD card code already contains code to handle the setting of the chip select line, there should be no need to do this within the HAL code. The thing is that in some circumstances the HAL SPI routines will be used by Marlin to access more than just the SD card, they can also be used to to talk to various LCD displays, with this current code you will end up with the SD card chip select being set when Marlin is trying to talk to the LCD (which is probably not good). I'm pretty sure that the definition of the HAL for these routines is that they should not be setting the chip select pins at all. There is good reason for this, some devices require that you hold the chip select active while carrying out multiple SPI operations (I have a feeling that the SD card interface spec states this, though you may get away with it with some cards), you can't do this if you are setting/releasing it on every read/write call. Take a look at the following code and you will see that the Marlin SD code is already dealing with chip select...
https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/sd/Sd2Card.cpp#L158

**Comment by 3DSmitty** [comment]
@gloomyandy ok wow I never saw that Sd2Card already manipulates the SS_PIN for reading and writing to SD cards. Thanks for pointing that out. Now I am questioning this whole pull request... and figure out why it does not use that Sd2Card routine when SDSUPPORT is enabled and relies on the HAL to change the state of the SS_PIN. Is SOFTWARE_SPI enabled by default?

**Comment by gloomyandy** [comment]
Even with software SPI it will still handle the chip select pin in the Marlin code. 

One thing that is a little odd is that some of the other HALs (but not the LPC176x HAL) setup the SS_PIN as output in the spiBegin function, which seems a little strange. But perhaps that is required? To be honest I'm not totally sure what the rules are with Marlin for handling the initialisation of I/O pins. @p3p can you throw any light on this, any idea why other HALs set the SS_PIN to be output, but the LPC176x does not? Is any setup actually required when using the extDigitalWrite call (which is what the SD code uses to set the chip select pin)?

**Comment by p3p** [comment]
The chip select signal should always be initialised and controlled by the device driver, so it will be set to output by the sd card code. The other platforms are just really making sure its set to output.

https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/sd/Sd2Card.cpp#L245

**Comment by 3DSmitty** [comment]
@gloomyandy @p3p Ok then something strange is happening with PlatformIO it seems to be using the HAL and the ststm32 library instead of Sd2Card. With SDSUPPORT enabled it is not setting the SS_PIN low before a transaction and then back high afterwards. Also how do you tell Sd2Card which SPI Master device to use(SPI1 orSPI3)? Or better yet how can you enable multiple SPI Masters with Sd2Card and have more than one SD card. How does Marlin know which SD card to read write to?

**Comment by tpruvot** [comment]
Could you rebase (and squash) these 13 commits in one ? would be easier to get the changes

**Comment by thinkyhead** [comment]
I'm fairly certain that all SPI code runs in the same thread, so as long as `safe_delay` is not used within SPI read/write loops (which could invoke a different SPI operation) there should be no danger of SPI contention within Marlin itself. As long as Marlin doesn't try to steal the bus from some interrupt-driven device on the same bus, all should be well.

**Comment by thinkyhead** [comment]
> Could you rebase (and squash) these 13 commits in one ? would be easier to get the changes

@tpruvot — Here's the quickest way to get them…
```
git remote add -f 3DSmitty git@github.com:3DSmitty/Marlin.git
git checkout 3DSmitty/bugfix-2.0.x -b bf2_skr_mini_14287
```

File: Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h (hunk 1 above)
Comment (thinkyhead) [reply Marlin/src/pins/pins_BIGTREE_SKR_MINI_V1_1.h:137]: Take comments from random users with a grain of salt.

**Comment by thinkyhead** [comment]
> How does Marlin know which SD card to read write to?

Marlin recognizes one SD card for SD printing purposes. If the onboard SD card is being used for EEPROM separately from an external SD card being used for printing, then the internal SD card is **only** used for EEPROM emulation and it is accessed through a separate API.

**Comment by 3DSmitty** [comment]
@thinkyhead Thank you for the clarification and commit!

**Comment by thinkyhead** [comment]
Thank you, and everyone else for helping make sense of it all.

**Comment by p3p** [comment]
@thinkyhead These changes, updating the SS pin inside the HALs SPI functions, break the api as stated above. The caller, that is implementing the specific devices protocol, needs to control it for the duration of the communication "transaction" which may span many reads/writes.

**Comment by 3DSmitty** [comment]
@p3p what do you mean by "the caller"?

**Comment by tpruvot** [comment]
will do a commit on the "extra spi functions" not used for the sdcard... which should "fix" your problem/sentiment. See PR #14299

**Comment by gloomyandy** [comment]
Folks perhaps we should take a step back here and define exactly what Marlin needs from the HAL in terms of an SPI interface and how it is implemented by some of the other HALs. That might help with all of this. This is my understanding of things, it is mainly based on the LPC176x HAL...

The HAL needs to export a single set of functions that accesses a single (possibly shared) SPI bus, this set of functions are encapsulated in a single object called SPI. The pins used for this bus are defined by the following defines SCK_PIN, MISO_PIN, MOSI_PIN. This SPI bus is primarily used by Marlin to access the SD card used for printing. The handling of the chip select pin is the responsibility of the code calling this interface (typically the Marlin SD card code). However this interface can be used by other parts of Marlin if other devices are sharing the bus. In particular it may be used for access to the LCD (in the case of an LCD that uses the same SPI bus for the display and the SD card reader), or for access to say TMC2130 drivers, when using so called "Hardware SPI". 

The HAL can use either a hardware or software implementation for this SPI interface. In the case of the LPC176x HAL it will select to use a hardware implementation if the selected SCK_PIN, MISO_PIN, MOSI_PIN pins match one of the two hardware SPI devices available on that platform. The pins are also used to identify which of the two hardware devices to use. There is also a define (LPC_SOFTWARE_SPI) that will force the use of a software SPI implementation for this API.

So the above is my understanding of the HAL SPI interface as it stands today. However (unfortunately) this does not cover all of the SPI usage of Marlin... 

There are a number of additional SPI implementations used throughout the code. Typically the HAL interface defined above is the only one that uses a "hardware" implementation so often this is referred to as the "hardware SPI" and these other implementations are called "software SPI". Some of these additional implementations are semi public and in some cases are mapped to the above interface (in the case of a shared SPI bus, or because it is desired to gain access to "hardware SPI"). Examples of these are the SPI interfaces use by the TMCDriver library and the U8G LCD interface library. Both can use either the HAL provided "hardware SPI" or can use an additional "Software SPI". Because these implementations can end up being mapped to the HAL provided interface they typically follow the same rules for things like setting speed and the handling of chip select.

In addition to this there may be other "private" interfaces that are used to access specific devices. Often these only expect a single device to be attached to the particular bus and may not follow the above rules for chip select and other things. Typically these interfaces are software only. However in some cases the underlying framework or HAL code may use SPI to provide additional functionality and this may make use of the various hardware SPI devices available. A good example of this is the LPC176x code that makes the onboard card reader on many LPC176x boards available by USB. This functionality does not use the Marlin SD code and has a second SD card interface library within the LPC176x framework. This code always uses hardware SPI to access the reader. However it has been arranged that the HAL SPI interface can be mapped to this same SPI hardware device allowing the onboard SD reader to be "shared" by both the USB access code and by Marlin. In this case the "sharing" is at the hardware SPI level and care is taken to ensure that the reader is "owned" either by Marlin or by the USB code and the two never access the device at the same time (switching of ownership is perform by mounting/unmounting  the device M21/M22 operations). 

It may be that the eeprom device that you are implementing falls into the "provided by the HAL but using a private SPI device", but that depends if the eeprom will always be on its own SPI bus really. It may be that you can use the same base code (perhaps by having multiple instances) to provide both this "private" SPI bus and the public one (used by Marlin), but in that case you will need to be sure that things like read/write do not touch the chip select lines. 

Sorry this is so long, but Marlin's use of SPI is rather complex (you might say a bit of a mess), I think I've managed to capture most of the details. @p3p @thinkyhead please correct anything I've got wrong!

**Comment by 3DSmitty** [comment]
@p3p and @gloomyandy I understand what you are saying that in the SPI protocol the SS_PIN **CAN** be held low though multiple read / write operations, I really do. I have spent many hours researching and reading different examples of STM32 SPI. With that said I know the SPI protocol is pretty standard across platforms but the STM32 is a different beast than the LPC176x. With the implementation of PatformIO in Marlin 2.0 the STM32 relies heavily on an external library stm32duino / ststm32 for SPI functions. The stm32duino / ststm32 library just like the HAL_STM32F1 also does not handle the state of the SS_PIN. So if I am understanding you correctly the state of the SS_PIN should be handled outside of the HAL where do we put that? How should it be implemented? @xC0000005 do you have any thoughts on this with all your work on the Malyan boards?

**Comment by Grogyan** [comment]
As I am one that primarily uses SPI for temperature monitoring on the
Max31855, serious consideration into SPI priority.

As well as potentially 2 SD cards, the SD card for the firmware, and the SD
card for the remote display gcode execution

>

**Comment by 3DSmitty** [comment]
@Grogyan what STM32F1 board are you using to do all of this?

**Comment by thisiskeithb** [comment]
Related question: Will this interfere with TMC drivers that use SPI? BigTreeTech shows an SPI hardware config (no firmware, of course) for 2130 if you skip the LCD: [image](https://imgur.com/a/s6APzJ4).

Edit: Assuming we add the proper pins to the SKR Mini’s pins files to get it working, that is.

**Comment by Grogyan** [comment]
I have an SKRV1.3 enroute, and maybe plan to get the SKR 1.1 Pro for
potentially looking at multi material printing

>

**Comment by 3DSmitty** [comment]
@Grogyan the SKR V1.3 and SKR V1.1 are LPC176x and the SKR Pro is STM32F4.

**Comment by thisiskeithb** [comment]
The SKR Pro is has an STM32F429ZGT6 chip on it.

**Comment by 3DSmitty** [comment]
@thisiskeithb Nice! Looks like a STM32F4. As far as using SPI3 to communicate with the TMC drivers per the pic this should not be a problem. The pins_BIGTREE_SKR_MINI_V1_1.h file may need tweaking a bit.

**Comment by tpruvot** [comment]
so its another HAL..

**Comment by gloomyandy** [comment]
> So if I am understanding you correctly the state of the SS_PIN should be handled outside of the HAL where do we put that? How should it be implemented?

You don't need to put it anywhere. Marlin already contains the code to deal with the SS_PIN (which is only used when accessing the SD card), other Marlin code and the TMCStepper library etc. also contains code that may be used with the SPI interface when setting chip select for other devices on the same bus. 

Just to be clear there is nothing special about the "SS_PIN" it is simply the chip select pin used by Marlin for the SD card reader. Unless you are writing code to access the SD card you don't need to touch it. Are you writing code to access the SD card? If so what does it do?

**Comment by 3DSmitty** [comment]
@gloomyandy than that's where the problem is then. If I take the code handling the SS_PIN out of the HAL_STM32F1 (like it was before) then Marlin does NOT deal with the SS_PIN  and the SD is non functional. This pull request gives access to the SD card allows reads and writes. So then something is broken between Sd2Card, HAL_STM32F1, and the stm32duino / ststm32 library.

**Comment by gloomyandy** [comment]
In which case there are a few possibilities. Clearly the code to modify the SS_PIN exists in the Marlin SD card modules (I gave a link to it earlier), so some possible reasons that code is not working....
 * Does the environment you are using correctly implement extDigitalWrite (used by the sd card code to set/clear the SS_PIN)? It looks like it is simply defined as digitalWrite, I assume that should work?
 * Is SS_PIN defined correctly when the Marlin SD code is being compiled? Is there any chance that any of the headers etc. has ended up redefining it? Should be possible to add some code to check this against an absolute pin value, or simply change the SD card code to use an explicit pin for testing. Does the spi_pins 
* Some sort of build issue means the SD card code has not been re-built with new definitions?

I don't have the hardware available to test this so I'm afraid I can't help directly. I'm sure you are probably fed up with all of this by now but the current code is not going to work with any sort of shared SPI environment, so it would be worth getting it right (because even if this particular board does not use a shared SPI bus other users of this HAL may do so).

I'd also like to question the way that SS_PIN is being defined. At the moment the code in spi_pins.h is setting the value of SS_PIN based upon the choice of hardware SPI device. But is this correct? Will all boards that use this HAL always define the chip select pin for the SD card as the one defined in this header? On most boards pretty much any gpio pin can be used for chip select functions.

**Comment by 3DSmitty** [comment]
@thisiskeithb the BIGTREETECH SKR MINi E3 looks interesting... it has STM32F103RCT6 and TMC2208 onboard. Looks like it is shipping at the end of this month and has Marlin 2.0. I wonder if they will release their source code.

**Comment by 3DSmitty** [comment]
@gloomyandy is there a difference between extDigitalWrite and digitalWrite? I had the SS_PIN defined in the board pins file originally but it didn't seem to make a difference. It is now defined in the spi_pins.h file. The pin assignments in spi_pins.h are determined by what board is choosen in platformio.ini. If these pins were getting redefined I would see it as a warning message when building. SPI.h in the stm32duino / ststm32 library does reference a external SdFat library???

I am not fed up at all, I setup this pull request the best way I knew how and if it is not right I would like to make sure to get it right. Thank you for your help.

**Comment by gloomyandy** [comment]
Hmm the Marlin sd card code actually uses SDSS not SS_PIN directly. I wonder if you need to add something like the following to the end of your spi_pins.h file (this comes from the LPC1768 spi_pins.h file)....
```
#if !defined(SDSS) || SDSS == P_NC // gets defaulted in pins.h
  #undef SDSS
  #define SDSS              SS_PIN
#endif
```
Personally I think it would be best to have the SS_PIN defined in the board pins file. That way other boards can use the pins as they see fit, at the moment they would have to change the HAL which is not a good thing really.

**Comment by 3DSmitty** [comment]
@gloomyandy ok cool let me play around with it and I will post my results.

**Comment by 3DSmitty** [comment]
@gloomyandy also what code determines which SPI to use for SD? For example on this board the onboard SD is SPI1 and the external SD (on LCD) is SPI3. How can I switch it from SPI1 to SPI3?

**Comment by gloomyandy** [comment]
On the LPC176x boards we have settings in the pins files that define the various pins to use for the two SD card configurations, then in the configure_adv there is a setting which lets the user specify which card reader they want to use (along with some other options about sharing the SD card via USB). The code in the HAL then looks at the settings for SCK_PIN, MISO_PIN, MOSI_PIN and decides which hardware SPI unit to use. Though I think that your current mechanism of having a define in the pins file for defining which SPI to use is also fine. The advantage of specifying actual pins is that it allows for the use of software SPI if required. 

You can see the code that does this for the LPC176x here... https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/HAL/HAL_LPC1768/HAL_spi.cpp
It is certainly not perfect, but that's how it works at the moment on that platform.

**Comment by 3DSmitty** [comment]
Ok I got it! How should I proceed at his point? Create a new fork with the changes and open a new pull request?

Thanks for your help @gloomyandy !!!

**Comment by gloomyandy** [comment]
So does it work now? That's brilliant if it does! I think the old code was merged so, yes a new PR is probably best, but @thinkyhead is probably the best one to cover that!

**Comment by 3DSmitty** [comment]
yes it does :-) And it seems to read / write / print from sd alot faster!

**Comment by 3DSmitty** [comment]
I couldn't get it to work with the  
SPI.beginTransaction(spiConfig);
or
SPI.endTransaction();

But if the SPI device is already initialized with spiInit(); then there is really no need for them. Removing them and dressing up spi_pins.h makes it work. So that tells me the actual bug has to do with them and they are part of the  stm32duino / ststm32 library.

**Comment by 3DSmitty** [comment]
@gloomyandy you can see what i did here:
https://github.com/3DSmitty/Marlin_Bigtreetech_SKR_Mini_v1_1
THANK YOU FOR YOUR HELP!

**Comment by 3DSmitty** [comment]
New pull request created https://github.com/MarlinFirmware/Marlin/pull/14306

**Comment by tpruvot** [comment]
look like the SPI1 is the only channel declared by default, if 2 are required at the same time...

https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/libraries/SPI/src/SPI.cpp#L792

**Comment by peppekerstens** [comment]
Hello, just installed an BigTreeTech SKR V1.3 on my Creality Ender 5.  Build branche: Marlin 2.0.x
Printing via USB = ok

The SD Card is not working as expected:
- Detectable and addressable from/via PC/USB 
- Does not init via Marlin firmware/LCD interface (did NOT try console yet)
- Enabled #define SDSUPPORT in configuration.h

Trying to find a solution/cause or at least an open issue. Reading through this, this seems the cause or is at least related?

**Comment by tpruvot** [comment]
this one is not related... (to STM chips) except the brand name

**Comment by peppekerstens** [comment]
Thank you for quick response. I'll look further then :)

**Comment by 3DSmitty** [comment]
@tpruvot you can change spi channel by SPI.setModule(). There is no reason why you couldn't initialize and use more than one SPI channel in the Marlin HAL. The Marlin HAL code is only setup for SPI masters though it does not allow you to do SPI Slave.

**Comment by tpruvot** [comment]
yes, but.. in theory you could have 2 (or 3) objects to keep settings of the 2 channels

**Comment by 3DSmitty** [comment]
true

**Pull request merged**